internal/process/zombie_monitor.go

```diff
@@ -0,0 +1,79 @@
+package process
+
+import (
+	"context"
+	"sync/atomic"
+	"syscall"
+	"time"
+
+	"github.com/cri-o/cri-o/internal/log"
+	"github.com/pkg/errors"
+)
+
+// Unfortunately, the metrics setup requires
+// the processes are retrievable without accessing the server.
+// Instead of working around this, create a global singleton,
+// which will be initialized with the server.
+var monitor *ZombieMonitor
+
+// ZombieMonitor is a structure for watching and cleaning up zombies on the node.
+// It has two tasks, both related to detecting the existence of defunct processes:
+// 1: it is responsible for reporting the number of defunct PIDs currently on the node
+// 2: it is responsible for cleaning up zombies that are children of the currently running process
+// It does so by occasionally polling for the zombie processes.
+// If any zombies are found, there is a delay between when they're identified and when they're cleaned.
+// This is to ensure ZombieMonitor doesn't interfere with the go runtime's own child management.
+type ZombieMonitor struct {
+	zombieCount uint32
+	closeChan   chan struct{}
+}
+
+// InitializeZombieMonitor creates and starts the zombie monitor.
+// It fails if one has already been created.
+func InitializeZombieMonitor() error {
+	if monitor != nil {
+		return errors.New("zombie monitor already intialized")
+	}
+	monitor = &ZombieMonitor{
+		closeChan: make(chan struct{}, 1),
+	}
+	go monitor.Start()
+	return nil
+}
+
+// CloseZombieMonitor instructs the zombie monitor to stop listening and exit.
+func CloseZombieMonitor() {
+	monitor.closeChan <- struct{}{}
+	monitor = nil
+}
+
+// Start begins the zombie monitor. It will populate the zombie count,
+// as well as begin the zombie cleaning process.
+func (zm *ZombieMonitor) Start() {
+	for {
+		count, zombieChildren, err := ParseDefunctProcesses()
+		if err != nil {
+			log.Warnf(context.Background(), "Failed to get defunct process information: %v", err)
+		}
+		atomic.StoreUint32(&zm.zombieCount, uint32(count))
```

Comment:
Should this run after the `case <-time.After(time.Second * 5):` below?

Reply:
dropped now